### ISSUE TITLE
feat: implement SWAG for publication + real SWA/Multi-SWA/Multi-SWAG/PSWA as P2

### ISSUE THREAD

**Issue by petteriTeikari** [petteriTeikari/minivess-mlops#892]
## Decision

**SWAG is the publication-gate post-training method.** Checkpoint averaging is removed
from the factorial design (negligible effect, not a real algorithm for our use case).

## Literature report

See: `docs/planning/swa-swag-multi-swa-swag-pswa-lawa-model-soup-checkpoint-averaging-report.md`

## Current state

The codebase has "SWA" plugins that are actually **post-hoc checkpoint averaging** — loading
saved best-metric checkpoints and computing arithmetic mean. This is NOT SWA. The debug run
confirmed <0.001 DSC improvement because all 7 checkpoints come from the same training
trajectory (best_val_loss, best_val_dice, etc. from nearby epochs).

See: `.claude/metalearning/2026-03-21-fake-swa-checkpoint-averaging-mislabeled.md`

## Immediate actions

1. **Rename** all "SWA" references to "checkpoint averaging" (~18 files)
2. **Remove** checkpoint averaging from factorial factor design (not publication-worthy)
3. **Implement SWAG** as the real post-training method

## Publication-gate: SWAG (Maddox et al. 2019)

**Paper**: [A Simple Baseline for Bayesian Deep Learning](https://arxiv.org/abs/1902.02476)
**Code**: [github.com/wjmaddox/swa_gaussian](https://github.com/wjmaddox/swa_gaussian)

SWAG extends SWA by tracking first and second moments of weights during the SWA
collection phase, constructing a low-rank-plus-diagonal Gaussian posterior approximation.
At inference, sample K models from the posterior for calibrated uncertainty.

### Implementation plan

SWAG requires **resumed training** in the post-training flow:

1. Load trained checkpoint + training DataLoaders
2. Resume training with `SWALR` for N additional epochs
3. Track: running mean, running squared mean, low-rank deviation matrix
4. `update_bn(train_loader, swag_model)` after collection
5. Save SWAG model (mean + covariance parameters)
6. At Analysis flow inference: sample K models, average predictions

### Factorial design update

```yaml
post_training:
  method:
    - none     # Use training checkpoint as-is
    - swag     # Maddox et al. 2019
```

### Real SWA (Izmailov et al. 2018)
- [Averaging Weights Leads to Wider Optima](https://arxiv.org/abs/1803.05407)
- `torch.optim.swa_utils.AveragedModel` + `SWALR` + `update_bn()`
- PyTorch blog by Izmailov: "Even if you have already trained your model..."

### Multi-SWA / Multi-SWAG (Wilson & Izmailov 2020)
- [Bayesian Deep Learning and a Probabilistic Perspective of Generalization](https://arxiv.org/abs/2002.08791)
- [github.com/izmailovpavel/understandingbdl](https://github.com/izmailovpavel/understandingbdl)
- M independent SWA/SWAG models ensembled

### PSWA — Periodic SWA (Guo et al. 2022)
- [SWA Revisited](https://arxiv.org/abs/2201.00519)
- Apply SWA multiple times sequentially for incremental gains

### LAWA — Latest Weight Averaging (Kaddour 2022)
- [Stop Wasting My Time!](https://arxiv.org/abs/2209.14981)
- Rolling average of k latest checkpoints during training

### Model Soups (Wortsman et al. 2022)
- [Model Soups: Averaging Weights of Multiple Fine-Tuned Models](https://arxiv.org/abs/2203.05482)
- Average independently trained models (different hyperparameters)

### SeWA, TWA (2023-2025)
- Selective and trainable weight averaging variants
- [SeWA: arXiv:2502.10119](https://arxiv.org/abs/2502.10119)
- [TWA: arXiv:2205.13104](https://arxiv.org/abs/2205.13104)

## Checkpoint averaging (keep but rename)

What the current code does is valid as a utility — rename to `CheckpointAveragingPlugin`.
Keep it for quick debugging but NOT as a factorial factor. Create P2 issue for
tracking multiple best-metric checkpoints properly.

## References

- [Literature report](docs/planning/swa-swag-multi-swa-swag-pswa-lawa-model-soup-checkpoint-averaging-report.md)
- [Metalearning](`.claude/metalearning/2026-03-21-fake-swa-checkpoint-averaging-mislabeled.md`)
- [PyTorch SWA blog](https://pytorch.org/blog/pytorch-1-6-now-includes-stochastic-weight-averaging/)
- [PyTorch swa_utils docs](https://pytorch.org/docs/stable/optim.html#stochastic-weight-averaging)
- ["When, Where and Why to Average Weights?" (2025)](https://arxiv.org/abs/2502.06761)
